test/ziggurat/messaging/producer_test.clj

```diff
@@ -520,32 +522,76 @@
     (let [publish-called (atom 0)]
       (with-redefs [lch/open                (fn [_] (reify Channel (close [_] nil)))
                     lb/publish              (fn [_ _ _ _ _]
-                                              (when (< @publish-called 2)
+                                              (when (< @publish-called 10)
                                                 (swap! publish-called inc)
                                                 (throw (IOException. "io exception"))))
                     metrics/increment-count (fn [_ _ _] nil)]
         (producer/publish "random-exchange" {:topic-entity "hello"} 12345)
-        (is (= 2 @publish-called)))))
+        (is (= 10 @publish-called)))))
   (testing "publish/producer tries to publish again if already closed exception is received"
     (let [publish-called (atom 0)]
       (with-redefs [lch/open                (fn [^Connection _] (reify Channel (close [_] nil)))
                     lb/publish              (fn [_ _ _ _ _]
-                                              (when (< @publish-called 2)
+                                              (when (< @publish-called 10)
                                                 (swap! publish-called inc)
                                                 (throw (AlreadyClosedException. (ShutdownSignalException. true true nil nil)))))
                     metrics/increment-count (fn [_ _ _] nil)]
         (producer/publish "random-exchange" {:topic-entity "hello"} 12345)
-        (is (= 2 @publish-called)))))
-  (testing "producer/publish does not try again if the exception thrown is neither IOException nor AlreadyClosedException"
+        (is (= 10 @publish-called)))))
+  (testing "publish/producer tries to publish again if TimeoutException is received"
     (let [publish-called (atom 0)]
       (with-redefs [lch/open                (fn [^Connection _] (reify Channel (close [_] nil)))
                     lb/publish              (fn [_ _ _ _ _]
-                                              (when (< @publish-called 2)
+                                              (when (< @publish-called 10)
+                                                (swap! publish-called inc)
+                                                (throw (TimeoutException. "timeout"))))
+                    metrics/increment-count (fn [_ _ _] nil)]
+        (producer/publish "random-exchange" {:topic-entity "hello"} 12345)
+        (is (= 10 @publish-called)))))
+  (testing "producer/publish tries again the number of times defined in the config if the exception thrown is non recoverable and if retry is enabled"
```

Comment:
[needs rephrasing] Example: "producer/publish retries publishing for a certain number of times (configurable) when a non-recoverable exception is thrown"